libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts

```diff
@@ -129,7 +115,10 @@ export class AsyncOperationQueue
         // This operation is not yet ready to be executed
         // next one plz :)
         continue;
-      } else if (record.status === OperationStatus.RemoteExecuting) {
+      } else if (
+        record.status === OperationStatus.RemoteExecuting ||
+        record.status === OperationStatus.RemoteExecutingPossiblyComplete
```

Comment:
I confess I hadn't realized that this was the cobuild _completion_ detector; my read of the initial implementation of the returning a RemoteExecuting operation had been that the runner was coded to execute the operation locally if it waiting longer than a certain delay with nothing to do.

As a result, we can simplify the logic a lot by just returning operations with status `RemoteExecutingPossiblyComplete` as ordinary candidates for execution. We can completely remove the `tryGetRemoteExecutingOperation` function, since as soon as operations in stathe `RemoteExecutingPossiblyComplete` get handed to the runner, their status will potentially change back to `RemoteExecuting`.

Basically the flow will be:
Ready -> Queued -> RemoteExecuting -(wait 500)-> RemoteExecutingPossiblyComplete -> Queued -> RemoteExecuting or terminal state

Reply:
Alternatively, we could just reset the operation to `Ready` instead of having the separate `RemoteExecutingPossiblyComplete` status. If we keep the separate status, line 125 should be updated to include it as one of the possible states for an operation being returned from the primary loop.

Reply:
I tried this, but ended up getting sidetracked trying to remove `RemoteExecuting` as a whole concept. All `RemoteExecuting` seems to be doing is allowing the queue to not mark an operation as complete and allow it to be requeued. Ideally, that would be possible if you return _any_ non-terminal status. If I return `Waiting` or `Executing` from a `beforeExecuteOperation` step, I think I'd want my operation to be requeued, not for the process to throw an error (as it does now). I ended up running into a few more issues trying to untangle the `RemoteExecuting` state and abandoned it, but I'd be curious to get your feedback on that direction?

Basically, we'd remove `RemoteExecuting`, `CacheableOperationPlugin` would return `Waiting` or `Executing` if it was unable to get the lock. `CacheOperationPlugin` would also set a timeout to set that operation's status to `Ready` after 500ms for it to be rechecked. The `onOperationStatusChanged` would be as you suggestion, only when the change is from x -> `Ready` do we trigger a call to `assignOperations`.

Reply:
Updated to the above model in [7b7e7f1](https://github.com/microsoft/rushstack/pull/4821/commits/7b7e7f170478e5b55b8af3a3b4bd622953f8771f). Tested it a bit locally and didn't run into any issues.